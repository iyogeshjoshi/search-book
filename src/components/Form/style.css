.search-form {
  margin-bottom: 2vmin;
}

.search-form #query {
  padding: 1vmin;
  font-size: calc(10px + 1vmin);
  border-radius: 5px;
  width: 95vmin;
  min-width: 200px;
}
