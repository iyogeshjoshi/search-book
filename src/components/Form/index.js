import React, { useRef, useState } from "react";
import "./style.css";

const Form = ({ books, searchHandler, selectionHandler }) => {
  const [quantity, setQuantity] = useState(3);
  const queryRef = useRef(null);

  let timeout;
  const onSearch = (e) => {
    const targetValue = e.target.value;
    timeout && clearTimeout(timeout);
    timeout = setTimeout(() => {
      targetValue && searchHandler(targetValue, quantity);
    }, 500);
  };
  const onOptionSelect = (book) => {
    queryRef.current.value = "";
    selectionHandler(book);
  };

  return (
    <>
      <section className="search-form">
        <input
          type="number"
          name="quantity"
          id="quantity"
          defaultValue={quantity}
          onChange={(e) => setQuantity(e.target.value)}
          min={1}
          title="number of results to show"
        />
        <input
          type="search"
          list="titles"
          name="query"
          id="query"
          placeholder="search here"
          title="Enter your input here to search books"
          tabIndex="1"
          onChange={onSearch}
          autoFocus
          ref={queryRef}
        />
        <span> Found: {books.length}</span>
        <div className={`dropdown ${books.length > 0 && "show"}`} id="titles">
          {books.map((book, index) => (
            <div
              key={`title-${index}`}
              className="option"
              onClick={() => onOptionSelect(book)}
            >
              {book.title}
            </div>
          ))}
        </div>
      </section>
    </>
  );
};

export default Form;
