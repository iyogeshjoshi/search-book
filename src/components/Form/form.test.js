import React from "react";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Form from "./";

describe("renders Form correctly", () => {
  const searchHandler = jest.fn();
  const selectionHandler = jest.fn();
  it("renders books", () => {
    const Books = [
      {
        title: "Sample title",
      },
    ];
    const { getByText } = render(
      <Form
        books={Books}
        searchHandler={searchHandler}
        selectionHandler={selectionHandler}
      />
    );

    expect(getByText(/Sample title/i)).toBeTruthy();
  });
  it("matches snapshot", () => {
    const Books = [];
    const form = renderer
      .create(
        <Form
          books={Books}
          searchHandler={searchHandler}
          selectionHandler={selectionHandler}
        />
      )
      .toJSON();

    expect(form).toMatchSnapshot();
  });
});
