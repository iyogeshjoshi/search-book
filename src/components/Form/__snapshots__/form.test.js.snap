// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Form correctly matches snapshot 1`] = `
<section
  className="search-form"
>
  <input
    defaultValue={3}
    id="quantity"
    min={1}
    name="quantity"
    onChange={[Function]}
    title="number of results to show"
    type="number"
  />
  <input
    autoFocus={true}
    id="query"
    list="titles"
    name="query"
    onChange={[Function]}
    placeholder="search here"
    tabIndex="1"
    title="Enter your input here to search books"
    type="search"
  />
  <div
    className="dropdown false"
    id="titles"
  />
</section>
`;
